src/github.com/matrix-org/go-neb/matrix/matrix.go

```diff
@@ -263,12 +261,54 @@ func (cli *Client) Sync() {
 		// a malformed/buggy event which keeps making us panic.
 		cli.NextBatchStorer.Save(cli.UserID, nextToken)
 
-		if !isFirstSync {
+		if processResponse {
+			// Update client state
 			channel <- syncResponse
 		}
 	}
 }
 
+// shouldProcessResponse returns true if the response should be processed. May modify the response to remove
+// stuff that shouldn't be processed.
+func (cli *Client) shouldProcessResponse(tokenOnSync string, syncResponse *syncHTTPResponse) bool {
+	if tokenOnSync == "" {
+		return false
+	}
+	// This is a horrible hack because /sync will return the most recent messages for a room
+	// as soon as you /join it. We do NOT want to process those events in that particular room
+	// because they may have already been processed (if you toggle the bot in/out of the room).
+	//
+	// Work around this by inspecting each room's timeline and seeing if an m.room.member event for us
+	// exists and is "join" and then discard processing that room entirely if so.
```

Comment:
I wonder if you'd want to process the !commands from after the last join event in the timeline.